sway-core/src/asm_generation/checks.rs

```diff
@@ -68,6 +75,115 @@ fn check_for_contract_opcodes(ops: &[AllocatedOp]) -> CompileResult<()> {
+/// If so, throw an error.
+///
+/// All contract opcodes are not allowed in predicates. Except for RVRT that can
+/// be used to abort the predicate. One example of disallowed code is as follows:
+/// ```ignore
+/// pub fn burn(amount: u64) {
+///   asm(r1: amount) {
+///     burn r1;
+///   }
+/// }
+/// ```
+///
+/// Jumping backwards is not allowed in predicates so JMP and JNE are not allowed and
+/// the function verifies that the immediate of JI, JNEI, JNZI is greater than the opcode offset.
+///
+/// See: https://fuellabs.github.io/fuel-specs/master/vm/index.html?highlight=predicate#predicate-verification
+fn check_predicate_opcodes(ops: &[AllocatedOp]) -> CompileResult<()> {
+    use AllocatedOpcode::*;
+    let default_span =
+        sway_types::span::Span::new("no span found for opcode".into(), 0, 1, None).unwrap();
+    let mut errors = vec![];
+
+    for (op, opcode_index) in ops.iter().zip(0u32..) {
+        macro_rules! invalid_opcode {
+            ($name_str:literal) => {{
+                errors.push(CompileError::InvalidOpcodeFromPredicate {
+                    opcode: $name_str.to_string(),
+                    span: op
+                        .owning_span
+                        .clone()
+                        .unwrap_or_else(|| default_span.clone()),
+                });
+            }};
+        }
+        macro_rules! invalid_backward_jump {
+            ($name_str:literal) => {{
+                errors.push(CompileError::InvalidBackwardJumpFromPredicate {
+                    opcode: $name_str.to_string(),
+                    span: op
+                        .owning_span
+                        .clone()
+                        .unwrap_or_else(|| default_span.clone()),
+                });
+            }};
+        }
+        match op.opcode.clone() {
+            BAL(..) => invalid_opcode!("BAL"),
+            BHEI(..) => invalid_opcode!("BHEI"),
+            BHSH(..) => invalid_opcode!("BHSH"),
+            BURN(..) => invalid_opcode!("BURN"),
+            CALL(..) => invalid_opcode!("CALL"),
+            CB(..) => invalid_opcode!("CB"),
+            CCP(..) => invalid_opcode!("CCP"),
+            CROO(..) => invalid_opcode!("CROO"),
+            CSIZ(..) => invalid_opcode!("CSIZ"),
+            GM(_, VirtualImmediate18 { value: 1..=2 }) => {
+                errors.push(CompileError::GMFromExternalContract {
+                    span: op
+                        .owning_span
+                        .clone()
+                        .unwrap_or_else(|| default_span.clone()),
+                });
+            }
+            JI(imm) => {
+                if imm.value <= opcode_index {
+                    invalid_backward_jump!("JI");
+                }
+            }
+            JMP(..) => invalid_opcode!("JMP"),
+            JNE(..) => invalid_opcode!("JNE"),
+            JNEI(_, _, imm) => {
+                if u32::from(imm.value) <= opcode_index {
```

Comment:
Is index of the `opcode` in the list `ops` the same as what jumps instructions use? Maybe that's true for `AllocatedOp` but I'll let @otrho confirm.

Reply:
Yeah, nah, it's an address at this stage.

The `VirtualOpcode`s and `AllocatedOpcode`s have addresses as args, though I don't think we use the `VirtualOpcode` ones.  Prior to address realisation we have `ControlFlowOp` which does jumping to symbolic labels.  So we're never jumping to an index in an array.

To catch backwards jumps you'll need to know the address of the jump instruction.

Reply:
@otrho what do you think of changing [map_label_offsets](https://github.com/FuelLabs/sway/blob/a781aa4d990472cf664dc4c5a330988bbb0aa237/sway-core/src/asm_generation/allocated_abstract_instruction_set.rs#L307) to also set a field with the address of opcodes. The new field `address` would be in `AllocatedAbstractOp`, `RealizedOp` and `AllocatedOp`.

The other option we have is to compute address in place but this would duplicate the places we handle with offsets.

Reply:
It only makes sense to perhaps add the address to `RealizedOp`, not the others.

But I've taken a closer look, and at the point where you're making these checks you have a fully formed binary blob which is basically an array of 32bit instructions, so calculating the address of each one is trivial -- not that different to what you're doing now, as you should be able to take the index of the instruction you're checking and just mul by 4.  I think?

Reply:
In the opcode docs it says for [JI](https://fuellabs.github.io/fuel-specs/master/vm/instruction_set.html#ji-jump-immediate) that `$pc = $is + imm * 4;` imm is multiplied by 4 to update the address so maybe we can compare imm with opcode_index directly.

Reply:
Oh, actually, I think you're right!  So at this stage, when everything is fully baked, the address _is_ the index and the code was OK all along.  Sorry.  It _might_ be more useful to rename `opcode_index` to `opcode_addr` just to document this.